ENH/FIX: use scipy for PSDs

yousrabk opened the pull request:
Back to the PR #1743... I've done all the changes you suggested @dengemann and @agramfort 
I still have some questions though and may be ways to do better.

agramfort commented on mne/epochs.py:
```diff
@@ -506,6 +507,51 @@ def plot(self, epoch_idx=None, picks=None, scalings=None,
+            Length of each window.
+        n_overlap : int
+            The number of points of overlap between blocks. The default value
+            is 0 (no overlap).
```
doctring is wrong. Does 1 make sense? I would have used window_size // 2

Eric89GXL replied:
+1 for allowing `'half'` which the code translates to `window_size // 2`

yousrabk replied:
Ok for window_size // 2

agramfort commented on mne/epochs.py:
```diff
@@ -38,6 +38,7 @@
 from .event import _read_events_fif
 from .fixes import in1d
 from .viz import _mutable_defaults, plot_epochs, _drop_log_stats
+from .viz.epochs import plot_psds
```
call this function plot_epochs_psds or maybe better plot_epochs_psd

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -34,16 +35,13 @@ def compute_raw_psd(raw, tmin=0., tmax=np.inf, picks=None,
-        performing the FFT. If None, pad_to equals `n_fft`.
     n_overlap : int
         The number of points of overlap between blocks. The default value
         is 0 (no overlap).
```
docstring is not uptodate

agramfort commented:
what are the timings with n_jobs=2 or more?

Eric89GXL commented:
Can you add a test that compares your computation to the output of `mpl.psd` to confirm that it hasn't changed? There might be normalization factors that one implementation uses that another does not.

yousrabk commented:
@Eric89GXL Yes sure I will add it.

yousrabk commented:
@Eric89GXL I was playing with the parameters to compare the results with mpl.psd. We have exactly the same figures and same normalization factors with window_size=2048 (too long?) and n_overlap=0.

If I keep the n_overlap by default as half window_size (1024), the normalization factors, the sharpness... change but I think as it is due to the parameters we choose, it should be okay, isn't it?

This figure is with my computation using by_default paramters: 
window_size=1024 and n_overlap=512
![raw_psds_3_w1024_nol512](https://cloud.githubusercontent.com/assets/4352445/5815772/c64d6da0-a066-11e4-87f7-fdba2ea7ade9.png)

And this one is with the other implementation using mpl.psd
![raw_psds_3_mpl_psd](https://cloud.githubusercontent.com/assets/4352445/5815773/c64e87c6-a066-11e4-9a78-28417895c765.png)

agramfort commented:
you need to add a proper test using an assert_*** function. Not just
visually.
You can run it a small simulated data if it makes testing easier/faster

agramfort commented:
also check the tests. Travis is not happy.

Eric89GXL commented:
Can you figure out what is driving the differences between the PSDs? You might have to go down to the level of the source code if the docstrings aren't sufficient to figure out how the computations differ. This seems like a minor point, but we don't want to switch normalization factors, etc. on people, so we want to make them equivalent if we can. If we can't, we should at least know why -- it'll come in handy in the future when we need to fix something.

yousrabk commented:
Hello all, sorry for the late commit. Here I fixed what was left and added the test to check that the two computation are giving the same results for psds and freqs.

agramfort commented on examples/time_frequency/plot_compute_epochs_data_spectrum.py:
```diff
@@ -0,0 +1,55 @@
+
+This script shows how to compute the power spectral density (PSD)
+of measurements on a epochs dataset. It also show the effect of applying SSP
+to the data to reduce ECG and EOG artifacts.
```
are you really playing with proj?

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -11,9 +12,9 @@
-def compute_raw_psd(raw, tmin=0., tmax=np.inf, picks=None,
-                    fmin=0, fmax=np.inf, n_fft=2048, pad_to=None, n_overlap=0,
-                    n_jobs=1, plot=False, proj=False, verbose=None):
+def compute_raw_psd(raw, tmin=0., tmax=np.inf, fmin=0, fmax=np.inf,
```
you should avoid changing parameters order. it will break people's code which means yours in 6 months ;) put back picks where is was and check other params.

agramfort replied:
this is not addressed

yousrabk replied:
Oops... I really thought it was tmin, tmax, fmin, fmax everywhere. Will fixed it

agramfort replied:
still needs fix

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -146,15 +151,37 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
+                                                verbose=verbose)
+
+    psds = np.empty(epochs.get_data()[:, picks].shape[:-1] +
+                    (n_fft // 2 + 1,))
```
you should call only once to avoid reading twice from disk

agramfort commented on mne/time_frequency/tests/test_psd.py:
```diff
@@ -82,3 +96,55 @@ def test_psd_epochs():
     assert_true(psds.shape == (1, len(picks), len(freqs)))
     assert_true(np.sum(freqs < 0) == 0)
     assert_true(np.sum(psds < 0) == 0)
+
```
pep8 2 lines between functions

agramfort commented on mne/time_frequency/tests/test_psd.py:
```diff
@@ -82,3 +96,55 @@ def test_psd_epochs():
+    picks = pick_types(raw.info, meg='grad', eeg=False, stim=False,
+                            exclude=exclude)
+
+    picks = picks[:2]
```
you can do it in one line

+    picks = pick_types(raw.info, meg='grad', eeg=False, stim=False,
+                            exclude=exclude)[:2]

agramfort commented on mne/time_frequency/tests/test_psd.py:
```diff
@@ -82,3 +96,55 @@ def test_psd_epochs():
+    freqs_mpl = freqs_mpl[mask]
+    psds_mpl = psds_mpl[:, mask]
+
+    assert_array_almost_equal(psds_welch, psds_mpl)
```
you need to add assert_array_almost_equal for freqs

agramfort commented on mne/time_frequency/tests/test_psd.py:
```diff
@@ -82,3 +96,55 @@ def test_psd_epochs():
+
+
+
+
```
remove empty lines. Keep only 1

agramfort commented on mne/viz/epochs.py:
```diff
@@ -447,3 +448,123 @@ def plot_epochs(epochs, epoch_idx=None, picks=None, scalings=None,
+                                         n_jobs=n_jobs)
+
+        # Convert PSDs to dB
+        psds = 10 * np.log10(psds)
```
dB should be 20 * log10

agramfort commented:
that's it for me now

yousrabk commented:
@agramfort here another commit if you can look at it again.
Thanks

agramfort commented:
can you provide time results and show results of the example?

yousrabk commented:
The time results for the raw.plot_psds(tmax=10.) is 711ms. (for me)
![psds_raw](https://cloud.githubusercontent.com/assets/4352445/5917228/ab31388a-a61a-11e4-9b45-47d8a1201dab.png)
![psd_raw_4ltmag](https://cloud.githubusercontent.com/assets/4352445/5917230/afd42456-a61a-11e4-8f6a-0a955953a470.png)

The time results for the epochs.plot_psds() is 11.3s. For only 4 picks: the 4 left temporal mag, the time is 703ms.
![psd_epochs](https://cloud.githubusercontent.com/assets/4352445/5917278/3c8a6ed2-a61b-11e4-940c-8f5f66bcdb9f.png)
![psd_epochs_4ltmag](https://cloud.githubusercontent.com/assets/4352445/5917281/4436e9f8-a61b-11e4-8300-81d404e4b8ef.png)

For n_jobs=4, the time goes from 11.3 to 9.4s

dengemann commented:
Let me check later and add some benchmarks.

2015-01-27 12:02 GMT+01:00 yousrabk <notifications@github.com>:

> The time results for the raw.plot_psds(tmax=10.) is 711ms. (for me)
> [image: psds_raw]
> <https://cloud.githubusercontent.com/assets/4352445/5917228/ab31388a-a61a-11e4-9b45-47d8a1201dab.png>
> [image: psd_raw_4ltmag]
> <https://cloud.githubusercontent.com/assets/4352445/5917230/afd42456-a61a-11e4-8f6a-0a955953a470.png>
>
> The time results for the epochs.plot_psds() is 11.3s. For only 4 picks:
> the 4 left temporal mag, the time is 703ms.
> [image: psd_epochs]
> <https://cloud.githubusercontent.com/assets/4352445/5917278/3c8a6ed2-a61b-11e4-940c-8f5f66bcdb9f.png>
> [image: psd_epochs_4ltmag]
> <https://cloud.githubusercontent.com/assets/4352445/5917281/4436e9f8-a61b-11e4-8300-81d404e4b8ef.png>
>
> For n_jobs=4, the time goes from 11.3 to 9.4s
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1748#issuecomment-71628260>.
>

yousrabk commented:
Yes please, thank you @dengemann :)

dengemann commented:
No, thanks for your efforts @yousrabk!

2015-01-27 12:07 GMT+01:00 yousrabk <notifications@github.com>:

> Yes please, thank you @dengemann <https://github.com/dengemann> :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1748#issuecomment-71629172>.
>

yousrabk commented:
> No, thanks for your efforts @yousrabk!
>
> No problem, It's a good learning experiment and a way to get myself back
to mne-python!


> 2015-01-27 12:07 GMT+01:00 yousrabk <notifications@github.com>:
>
> > Yes please, thank you @dengemann <https://github.com/dengemann> :)
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/mne-tools/mne-python/pull/1748#issuecomment-71629172>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1748#issuecomment-71630342>.
>



-- 
Yousra BEKHTI

agramfort commented:
the speed improvement with n_jobs is not great. We should be able to do
better. Taking a look now.

agramfort commented on examples/time_frequency/plot_compute_epochs_data_spectrum.py:
```diff
@@ -0,0 +1,53 @@
+==================================================
+
+This script shows how to compute the power spectral density (PSD)
+of measurements on a epochs dataset. 
```
on a epochs dataset.  -> on epochs

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -75,20 +73,27 @@ def compute_raw_psd(raw, tmin=0., tmax=np.inf, picks=None,
+
+    freq_mask = (freqs >= fmin) & (freqs <= fmax)
+    freqs = freqs[freq_mask]
+    psds = psds[:, freq_mask]
```
you could extract freq mask in my_pwelch to avoid unecessary allocations.

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -103,30 +108,32 @@ def _compute_psd(data, fmin, fmax, Fs, n_fft, psd, n_overlap, pad_to):
 @verbose
-def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
-                       pad_to=None, n_overlap=0, n_jobs=1, verbose=None):
+def compute_epochs_psd(epochs, fmin=0, fmax=np.inf, proj=False, n_fft=2048,
```
you changed API here too

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -146,15 +153,37 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
+                psds[i_iter * i_epoch + i_channel, :, :] = fchannel[1]
+        i_iter = len(fepoch)
+
+    freq_mask = (freqs >= fmin) & (freqs <= fmax)
```
same remark here you can mask in the loop

agramfort commented on mne/time_frequency/psd.py:
```diff
@@ -146,15 +153,37 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
+            if i_epoch == 0:
+                psds[len(fepoch) * i_epoch + i_channel, :, :] = fchannel[1]
+            else:
+                psds[i_iter * i_epoch + i_channel, :, :] = fchannel[1]
```
I don't get why you need to branch with i_epoch==0

I am sure you can avoid it.

yousrabk replied:
I've done it that way, because fepoch can have different sizes, and I needed the size of the previous iteration in order to have the right index in psds. 

e.g if I split my data of 100 epochs to 3 sets, the last set will have 34 epochs, so len(fepoch) won't be the same than previous to have the continuous index... Am I clear?

I'm looking again at it to see how can I avoid it.

agramfort commented on mne/viz/epochs.py:
```diff
@@ -447,3 +448,122 @@ def plot_epochs(epochs, epoch_idx=None, picks=None, scalings=None,
+                ax_list.append(plt.subplot(len(picks_list), 1, ii + 1))
+        make_label = True
+    else:
+        fig = ax_list[0].get_figure()
```
if this is copy pasted from plot_raw_psds you should factorize this code.

yousrabk replied:
What do you mean?

dengemann replied:
@yousrabk what @agramfort suggests is that we should avoid using copied and pasted code. Instead we could add a hidden, i.e., private function (name starts with an underscore) that is called from inside plot_raw_psds and from inside plot_epochs_psds.

agramfort replied:
+10

agramfort commented on mne/viz/raw.py:
```diff
@@ -665,12 +671,14 @@ def plot_raw_psds(raw, tmin=0.0, tmax=60.0, fmin=0, fmax=np.inf,
 
         # Convert PSDs to dB
-        psds = 10 * np.log10(psds)
+        psds = 20 * np.log10(psds)
```
oups my bad if psds is power (not amplitude) than it should be 10 * np.log10

please revert here and change above.

yousrabk commented:
@dengemann can you have a look on this?

dengemann commented on examples/time_frequency/plot_compute_epochs_data_spectrum.py:
```diff
@@ -0,0 +1,53 @@
+                 n_jobs=1, color=(0, 0, 1),  picks=picks)
+
+ax.set_title('Four left-temporal magnetometers')
+plt.show()
```
This adds another example .... we already have `examples/time_frequency/plot_single_trial_spectra.py` can we please avoid that and merge both examples together?

dengemann replied:
Perfomance looks excellent over here. Takes 1.33 seconds on the epochs with n_jobs=8 and default params. It outperforms my intial solution. For raw 1.5 seconds with defaults. Really cool!

agramfort replied:
+1 for merging examples

dengemann commented on mne/time_frequency/psd.py:
```diff
@@ -31,21 +32,18 @@ def compute_raw_psd(raw, tmin=0., tmax=np.inf, picks=None,
     fmax : float
         Max frequency of interest
+    proj : bool
+        Apply SSP projection vectors
```
...n vectors.

dengemann commented on mne/time_frequency/psd.py:
```diff
@@ -31,21 +32,18 @@ def compute_raw_psd(raw, tmin=0., tmax=np.inf, picks=None,
-        The number of points to which the data segment is padded when
-        performing the FFT. If None, pad_to equals `n_fft`.
+    window_size : int, optional
+        Length of each window.
```
mention the default value

dengemann commented on mne/time_frequency/psd.py:
```diff
@@ -75,20 +73,25 @@ def compute_raw_psd(raw, tmin=0., tmax=np.inf, picks=None,
 
-    return psd, freqs
+def _pwelch(epoch, window_size, noverlap, nfft, fs, freq_mask):
+    return [welch(channel, nperseg=window_size, noverlap=noverlap,
```
"""Aux function"""   pseudo-doc string

dengemann commented on mne/time_frequency/psd.py:
```diff
@@ -120,13 +124,14 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
-        is 0 (no overlap).
+        is 128 (window_size // 2).
+    proj : bool
+        Apply SSP projection vectors
```
.

dengemann replied:
defaults to ....

dengemann commented on mne/time_frequency/psd.py:
```diff
@@ -146,15 +151,30 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, n_fft=256,
+                                    freq_mask=freq_mask)
+                           for epoch in np.array_split(data, n_jobs))):
+        for k, p in zip(idx, fepoch):
+            psds[k, :, :] = p
```
can we use more explicit identifiers here? i_epochs, epochs... or i_channl, f_channel, etc.

dengemann commented on mne/viz/epochs.py:
```diff
@@ -447,3 +448,123 @@ def plot_epochs(epochs, epoch_idx=None, picks=None, scalings=None,
+    area_alpha : float
+        Alpha for the area.
+    window_size : int, optional
+        Length of each window.
```
here and above default values are not mentioned

dengemann commented on mne/viz/epochs.py:
```diff
@@ -447,3 +448,123 @@ def plot_epochs(epochs, epoch_idx=None, picks=None, scalings=None,
+    if make_label:
+        tight_layout(pad=0.1, h_pad=0.1, w_pad=0.1, fig=fig)
+    plt.show()
+    return fig
```
Maybe now or later we should have method raw/epochs.plot_psds_topomap to get a figure like here:

http://martinos.org/mne/stable/auto_examples/time_frequency/plot_single_trial_spectra.html

Although with a better vmin / vmax handling, maybe normalization (dividing each band by sum over frequencies).

dengemann commented on mne/viz/raw.py:
```diff
@@ -618,6 +619,11 @@ def plot_raw_psds(raw, tmin=0.0, tmax=60.0, fmin=0, fmax=np.inf,
     area_alpha : float
         Alpha for the area.
+    window_size : int, optional
+        Length of each window.
```
default value is not mentioned.

dengemann commented:
@yousrabk that's it for me for now. Nice work!

yousrabk commented:
Thank you @dengemann for the comments. I'm on it

yousrabk commented:
@dengemann I merged the examples by adding a new parameter to choose the kind of plot we want for epochs only. 

So in the example, we have epochs.plot_psds(plot_kind=1) (the new epoch_psds fig above) or epochs.plot_psds(plot_kind=2) (yours from the plot_single_trial_spectra.py e.g). The next one (in progress) would be the topomap plot.

agramfort commented:
plot_kind=1 is not explicit and plot_ is not necessary. Use something
like kind="spectrum"
and kind="raster" or "image"

yousrabk commented:
Ok

agramfort commented on examples/time_frequency/plot_single_trial_spectra.py:
```diff
@@ -85,15 +58,3 @@
-evoked.plot_topomap(ch_type='grad', times=range(5, 12, 2),
-                    scale=1, scale_time=1, time_format='%0.1f Hz',
-                    cmap='Reds', vmin=np.min, vmax=np.max,
-                    unit='dB', format='-%0.1f')
```
you'll need to add the plot_psds_topomap to keep this image in the doc

agramfort commented:
besides looks great. I like red lines :)

dengemann commented on examples/time_frequency/plot_single_trial_spectra.py:
```diff
@@ -35,48 +35,21 @@
 
-fig.suptitle('Single trial power', fontsize=12)
+# Let's first check out all channel types by averaging across epochs.
+epochs.plot_psds(plot_kind=1, fmin=2, fmax=200, n_fft=n_fft, n_jobs=2)
```
would be more consistent with how we handle things + more explicit to have different name here instead of a type/mode argument, why not epochs.plot_psds_image/raster ?

yousrabk replied:
Then, we would have epochs.plot_psds_spectrum, epochs.plot_psds_image and epochs.plot_psds_topomap?

As the two first ones are "almost" the same, we only thought about a parameter but I mean why not. So we go for three different functions?

agramfort replied:
I suggested the kind param to have pandas like api...

dengemann replied:
@yousrabk the one would show an image, the other an averaged spectrogram plot, right? Not quite the same. 
Thinking about it maybe let's drop the image plot. Coverage is already good with plot_epochs_psds and plot_epochs_psds_topomap. Let's rathet tweak the topomap function to make it as nice as as possible.

Consider this example:

```Python
def plot_topomap_multi_cbar(data, pos, ax, title=None, unit=None,
                            vmin=None, vmax=None,
                            cmap='RdBu_r', colorbar=True):
    """Low level plot topography for a single measure

    Parameters
    ----------
    data : numpy.ndarray of float, shape (n_sensors,)
        The data show.
    pos : numpy.ndarray of float, shape (n_sensors, 2)
        The positions of the sensors.
    ax : instance of Axis
        The axes to plot on.
    title : str | None
        The axes title to show. Defaults to None (no title will be shown).
    
"""
    import matplotlib.pyplot as plt
    from mpl_toolkits.axes_grid1 import make_axes_locatable

    ax.set_yticks([])
    ax.set_xticks([])
    ax.set_frame_on(False)
    vmin = np.min(data) if vmin is None else vmin
    vmax = np.max(data) if vmax is None else vmax
    if title is not None:
        ax.set_title(title)
    im, _ = plot_topomap(data, pos, vmin=vmin,
                         vmax=vmax,
                         axis=ax,
                         cmap=cmap,
                         image_interp='bilinear',
                         contours=False)
    if colorbar is True:
        divider = make_axes_locatable(ax)
        cax = divider.append_axes("right", size="5%", pad=0.05)
        cbar = plt.colorbar(im, cax=cax, ticks=(vmin, vmax),
                            format='%3.3f')
        if unit is not None:
            cbar.set_label(unit)
        cbar.ax.tick_params(labelsize=8)

```

dengemann replied:
... and then this:

```Python
And this:

```Python
def plot_spectral_power_maps(info, psds, freqs, bands=None, show=True,
                             agg_fun=np.sum, unit=None, cmap='jet'):
    """Plot topomaps of the power spectral densities.
    Creates 5 plots grouped by bands:
    0 - 4 Hz   (Delta)
    4 - 8 Hz   (Theta)
    8 - 12 Hz  (Alpha)
    12 - 30 Hz (Beta)
    30 - 45 Hz (Gamma)


    Parameters
    ----------
    info : instance of mne.meas_info.Info
        The measurement info.
    psds : np.ndarray of float, shape (n_channels, n_freqs)
        Power spectral densities
    freqs : np.ndarray of float, shape (n_freqs)
        Frequencies used to compute psds.
    bands : list of tuple | None
        The lower and upper frequency and the name for that band. If None,
        (default) expands to:

        bands = [(0, 4, 'Delta'), (4, 8, 'Theta'), (8, 12, 'Alpha'),
                 (12, 30, 'Beta'), (30, 45, 'Gamma')]

    show : bool
        Whether to show the figure or not. Defaults to True.
    agg_fun : callable
        The function used to aggregate over frequencies. Defaults to np.sum.

    Returns
    -------
    fig : instance of matplotlib.pyplot.Figure
        The figure object.
    """
    import matplotlib.pyplot as plt

    pos = mne.channels.make_eeg_layout(info, exclude=[]).pos
    if bands is None:
        bands = [(0, 4, 'Delta'), (4, 8, 'Theta'), (8, 12, 'Alpha'),
                 (12, 30, 'Beta'), (30, 45, 'Gamma')]

    n_axes = len(bands)
    fig, axes = plt.subplots(1, n_axes, figsize=(1.5 * n_axes, 2))
    if n_axes == 1:
        axes = [axes]

    for (fmin, fmax, title), ax in zip(bands, axes):
        freq_mask = (fmin < freqs) & (freqs < fmax)
        data = agg_fun(psds[:, freq_mask], axis=1)
        plot_topomap_multi_cbar(data, pos, ax, title=title,
                                unit=unit, cmap=cmap)

    mne.viz.tight_layout(fig=fig)
    fig.canvas.draw()
    if show is True:
        plt.show()
    return fig

```

dengemann replied:
It will create nice colorbars and allows for averaging togehter over frequencies.

agramfort replied:
do we want to add all these plot methods?

yousrabk commented:
Ok, I will remove the image plot then and add the topomap one asap.

yousrabk commented:
@dengemann, @agramfort plot_psds_topomap is here, if you can look at it.

This is the result of the example (btw may be we should change its name?):
![psd_epochs_topomap](https://cloud.githubusercontent.com/assets/4352445/5975240/6467ac98-a87e-11e4-9308-bc16f219fbfb.png)

agramfort commented on examples/time_frequency/plot_single_trial_spectra.py:
```diff
@@ -35,65 +35,18 @@
-                    baseline=(None, 0), reject=dict(grad=4000e-13, eog=150e-6))
-
+epochs = mne.Epochs(raw, events, event_id, tmin, tmax,
+					proj=True, baseline=(None, 0), preload=True,
```
fix indent

agramfort commented on examples/time_frequency/plot_single_trial_spectra.py:
```diff
@@ -35,65 +35,18 @@
-                    cmap='Reds', vmin=np.min, vmax=np.max,
-                    unit='dB', format='-%0.1f')
+# Now let's take a look at the spatial distributions of the psd.
+epochs.plot_psds_topomap(picks=picks, n_fft=n_fft, n_jobs=2)
```
use n_jobs=1 in examples.

agramfort commented on mne/epochs.py:
```diff
@@ -506,6 +507,111 @@ def plot(self, epoch_idx=None, picks=None, scalings=None,
+    def plot_psds_topomap(self, bands=None, vmin=None, vmax=None, proj=False,
+                          n_fft=2048, picks=None, window_size=256,
+                          n_overlap=128, layout=None, cmap='RdBu_r',
+                          agg_fun=np.sum, n_jobs=1, verbose=None):
```
this function is not tested

dengemann replied:
Also we need to approach the API a bit to the one we have in the AverageTFR object.. Second we should provie a 'normalize' flag (devide each frequency by sum of all frequencies).

agramfort replied:
>
> Also we need to approach the API a bit to the one we have in the
> AverageTFR object.. Second we should provie a 'normalize' flag (devide each
> frequency by sum of all frequencies).
>

you mean local detrending?

agramfort commented on examples/time_frequency/plot_single_trial_spectra.py:
```diff
@@ -1,11 +1,11 @@
-compute average spectra to identify channels and
-frequencies of interest for subsequent TFR analyses.
+This script shows how to compute the power spectral density (PSD)
+of measurements on epochs. It also show how to plot its spatial
```
show -> shows

also please run pep8 on this file.

agramfort commented on mne/epochs.py:
```diff
@@ -506,6 +507,111 @@ def plot(self, epoch_idx=None, picks=None, scalings=None,
+        verbose : bool, str, int, or None
+            If not None, override default verbose level (see mne.verbose).
+        """
+        return plot_epochs_psd(self, fmin=fmin, fmax=fmax, proj=proj,
```
sometimes we use psd sometimes psds. We should unify this. Can you have a look?

yousrabk replied:
Ok will keep only psd.

Eric89GXL commented:
@yousrabk do you have time to rebase this (again) and make the next round of changes? It would be good to get this into the next release. If you don't have time, someone else can take over.

yousrabk commented:
You can may be take a look at it? and @dengemann?
Thanks
